include/operations/SymbolicOperation.hpp

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include "Definitions.hpp"
+#include "Expression.hpp"
+#include "StandardOperation.hpp"
+
+#include <algorithm>
+#include <cstddef>
+#include <optional>
+#include <unordered_set>
+#include <variant>
+
+namespace qc {
+    // Overload pattern for std::visit
+    template<typename... Ts>
+    struct Overload: Ts... {
+        using Ts::operator()...;
+    };
+    template<class... Ts>
+    Overload(Ts...) -> Overload<Ts...>;
+
+    class SymbolicOperation final: public StandardOperation {
+    protected:
+        std::array<std::optional<Symbolic>, MAX_PARAMETERS> symbolicParameter{{std::nullopt, std::nullopt, std::nullopt}};
+
+        static OpType parseU3(const Symbolic& lambda, dd::fp& phi, dd::fp& theta);
+        static OpType parseU3(dd::fp& lambda, const Symbolic& phi, dd::fp& theta);
+        static OpType parseU3(dd::fp& lambda, dd::fp& phi, const Symbolic& theta);
+        static OpType parseU3(const Symbolic& lambda, const Symbolic& phi, dd::fp& theta);
+        static OpType parseU3(const Symbolic& lambda, dd::fp& phi, const Symbolic& theta);
+        static OpType parseU3(dd::fp& lambda, const Symbolic& phi, const Symbolic& theta);
```

Comment:
Again, I might be stupid, but isn't this missing a parsing method where all three parameters are symbolic?

Reply:
Not really, although I understand where the question comes from. The `parseU3` method is called when constructing a `SymbolicOperation` with the `U3` type. As far as I understand it, the `parseU3` method takes care of figuring out whether a U3 gate is truly a U3 gate and not just a, say, Hadamard gate. If all parameters are symbolic then no reduction could ever be possible.

Reply:
Yeah, agreed.